flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java

```diff
@@ -0,0 +1,2654 @@
+			@Override
+			public TriggerResult answer(InvocationOnMock invocation) throws Exception {
+				Trigger.TriggerContext context = (Trigger.TriggerContext) invocation.getArguments()[3];
+				// don't intefere with cleanup timers
+				timeAdaptor.registerTimer(context, 0L);
+				context.getPartitionedState(valueStateDescriptor).update("hello");
+				return TriggerResult.CONTINUE;
+			}
+		}).when(mockTrigger).onElement(Matchers.<Integer>anyObject(), anyLong(), anyTimeWindow(), anyTriggerContext());
+
+		doAnswer(new Answer<TriggerResult>() {
+			@Override
+			public TriggerResult answer(InvocationOnMock invocation) throws Exception {
+				Trigger.OnMergeContext context = (Trigger.OnMergeContext) invocation.getArguments()[1];
+				// don't intefere with cleanup timers
+				timeAdaptor.registerTimer(context, 0L);
+				context.getPartitionedState(valueStateDescriptor).update("hello");
+				return TriggerResult.CONTINUE;
+			}
+		}).when(mockTrigger).onMerge(anyTimeWindow(), anyOnMergeContext());
+
+		doAnswer(new Answer<Object>() {
+			@Override
+			public Object answer(InvocationOnMock invocation) throws Exception {
+				Trigger.TriggerContext context = (Trigger.TriggerContext) invocation.getArguments()[1];
+				timeAdaptor.deleteTimer(context, 0L);
+				context.getPartitionedState(valueStateDescriptor).clear();
+				return null;
+			}
+		}).when(mockTrigger).clear(anyTimeWindow(), anyTriggerContext());
+
+		when(mockAssigner.assignWindows(anyInt(), anyLong(), anyAssignerContext()))
+				.thenReturn(Arrays.asList(new TimeWindow(0, 2)));
+
+		testHarness.processElement(new StreamRecord<>(0, 0L));
+
+		assertEquals(3, testHarness.numKeyedStateEntries()); // window state plus trigger state plus merging window set
+		assertEquals(2, timeAdaptor.numTimers(testHarness)); // timer and GC timer
+
+		when(mockAssigner.assignWindows(anyInt(), anyLong(), anyAssignerContext()))
+				.thenReturn(Arrays.asList(new TimeWindow(2, 4)));
+
+		shouldMergeWindows(
+				mockAssigner,
+				Lists.newArrayList(new TimeWindow(0, 2), new TimeWindow(2, 4)),
+				Lists.newArrayList(new TimeWindow(0, 2), new TimeWindow(2, 4)),
+				new TimeWindow(0, 4));
+
+		// don't register a timer or update state in onElement, this checks
+		// whether onMerge does correctly set those things
+		doAnswer(new Answer<TriggerResult>() {
+			@Override
+			public TriggerResult answer(InvocationOnMock invocation) throws Exception {
+				return TriggerResult.CONTINUE;
+			}
+		}).when(mockTrigger).onElement(Matchers.<Integer>anyObject(), anyLong(), anyTimeWindow(), anyTriggerContext());
+
+		testHarness.processElement(new StreamRecord<>(0, 0L));
+
+		verify(mockTrigger).onMerge(eq(new TimeWindow(0, 4)), anyOnMergeContext());
+
+		assertEquals(3, testHarness.numKeyedStateEntries());
+		assertEquals(2, timeAdaptor.numTimers(testHarness));
+	}
+
+	@Test
+	public void testRejectShrinkingMergingEventTimeWindows() throws Exception {
+		testRejectShrinkingMergingWindows(new EventTimeAdaptor());
+
```

Comment:
private :) All of the testXXXXXXX branch methods for testing below can be private, I think.